src/Control/Monad/Fail/Hoist.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE CPP                    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+
+-- | 'HoistFail extends 'MonadFail with 'hoistFail, which enables lifting
+-- of partiality types such as 'Maybe' and @'Either' e@ into the monad.
+module Control.Monad.Fail.Hoist
+  ( HoistFail(..)
+  , hoistFail_
+  , hoistFailE
+  , hoistFailE_
+  , (<%!>)
+  , (<!>)
+  ) where
+
+import           Prelude hiding (fail)
+
+import           Control.Monad.Fail         (MonadFail (..))
+
+import           Data.Either                (Either, either)
+
+class Monad m => HoistFail t m e | t -> e where
+
+  -- | Given a conversion from the error in @t a@ to @String@, we can hoist the
+  -- computation into @m@.
+  --
+  -- @
+  -- 'hoistFail' :: 'MonadFail' m => (() -> e) -> 'Maybe'       a -> m a
+  -- 'hoistFail' :: 'MonadFail' m => (a  -> e) -> 'Either'  a   b -> m b
+  -- @
+  hoistFail
+    :: (e -> String)
+    -> t a
+    -> m a
+
+instance MonadFail m => HoistFail Maybe m () where
+  hoistFail f = maybe (fail $ f ()) pure
+
+instance MonadFail m => HoistFail (Either e) m e where
+  hoistFail f = either (fail . f) pure
+
+hoistFail_ :: HoistFail t m String => t a -> m a
+hoistFail_ = hoistFail id
```

Comment:
The `_` suffix usually means "discard result" to me; I'd probably call this `hoistFail'`.

Reply:
And maybe we want `hoistFailM`/`hoistFailM'`/`hoistFailEM`/`hoistFailEM'` for consistency with `hoistError` functions?

Reply:
I've added many functions.